File: Easy/3. Sorted Squared Array/sortedSequenceArray.go
package main

import (
	"fmt"
	"sort"
)

// O(nlogn) time | O(n) space
func solution1(array []int) []int {
	sortedArray := make([]int, len(array))

	for i, val := range array {
		sortedArray[i] = val * val
	}

	sort.Ints(sortedArray)
	return sortedArray
}

// O(n) time | O(n) space
func solution2(array []int) []int {
	sortedArray := make([]int, len(array))

	smaller := 0
	larger := len(array) - 1

	for i := len(array) - 1; i >= 0; i-- {
		smallerValue := array[smaller]
		largerValue := array[larger]

		if abs(smallerValue) > abs(largerValue) {
			sortedArray[i] = smallerValue * smallerValue
			smaller++
		} else {
			sortedArray[i] = largerValue * largerValue
			larger--
		}
	}
	return sortedArray
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func main() {
	array := []int{1, 2, 3, 5, 6, 8, 9}
	fmt.Println("Solution 1: ", solution1(array))
	fmt.Println("Solution 2: ", solution2(array))

}

File: Easy/3. Sorted Squared Array/sortedSequenceArray.py
# O(nlogn) time | O(n) space
def solution1(array):
    sortedArray = []
    for i in range(len(array)):
        sortedArray.append(array[i]**2)
    sortedArray.sort()
    return sortedArray

# O(n) time | O(n) space
def solution2(array):
    sortedArray = [0 for _ in array]
    smaller = 0
    larger = len(array) - 1

    for i in reversed(range(len(array))):
        smallerValue = array[smaller]
        largerValue = array[larger]
        
        if abs(smallerValue) > abs(largerValue):
            sortedArray[i] = smallerValue ** 2
            smaller += 1
        else:
            sortedArray[i] = largerValue ** 2
            larger -= 1

    return sortedArray

if __name__ == "__main__":
    array = [1, 2, 3, 5, 6, 8, 9]
    print(f"Solution 1: {solution1(array)}")
    print(f"Solution 2: {solution2(array)}")
